openshift/console
# CONSOLE-922: Support AppliedClusterResourceQuota for normal users

# List Pages

I added logic so the resource link knows what the current namespace is. We need to pass the user's current namespace to the resource link if it's an ACRQ because the ACRQ itself doesn't supply namespace metadata.

Scoped namespace:
<img width="1078" alt="Screen Shot 2021-09-24 at 5 33 14 PM" src="https://user-images.githubusercontent.com/7014965/134742491-45597258-8fb6-494f-bb0c-ca28775e0035.png">

All namespaces:
<img width="1072" alt="Screen Shot 2021-09-24 at 5 34 45 PM" src="https://user-images.githubusercontent.com/7014965/134742506-8cb083ab-fc29-4c7f-b579-a25fb08fb3cd.png">

Support for search page (links work):
<img width="1422" alt="Screen Shot 2021-09-28 at 1 14 30 PM" src="https://user-images.githubusercontent.com/7014965/135134565-2f38450c-e0f2-4e95-91c4-3bba2b0396e3.png">

Support for API explorer page (instances links work):
<img width="1427" alt="Screen Shot 2021-09-28 at 1 15 14 PM" src="https://user-images.githubusercontent.com/7014965/135134524-42c539de-2e3b-4904-bd5f-230b8684099c.png">

# Details Page

New details page:
<img width="1096" alt="Screen Shot 2021-09-28 at 4 13 56 PM" src="https://user-images.githubusercontent.com/7014965/136033181-7494787b-5947-4080-8976-11290e7c8321.png">

New total used field (total used is cluster scope and used is scoped to the current namespace):
<img width="1095" alt="Screen Shot 2021-09-28 at 1 07 06 PM" src="https://user-images.githubusercontent.com/7014965/135133301-43534ce7-2173-429c-9102-e502084fe4e4.png">

We show ACRQs in the Resource Quota card in the project overview:
<img width="549" alt="Screen Shot 2021-10-04 at 2 13 08 PM" src="https://user-images.githubusercontent.com/7014965/136032517-48a39bcd-b194-4e54-a66b-6c935841f65c.png">
<img width="544" alt="Screen Shot 2021-10-04 at 2 13 15 PM" src="https://user-images.githubusercontent.com/7014965/136032551-1b4ba746-5515-40a9-b684-defe54a9eb70.png">

I'm basing the meat of the list page part of the PR on some work Sam did a few years ago: https://github.com/spadgett/console/commits/applied-cluster-quota. 

Fixes https://issues.redhat.com/browse/CONSOLE-922.

# How to test:

1. Add new user (we are calling this user `rebecca` here) and login
2. Create a new project: `oc new-project ralpert-test`
3. Log in as admin
4. Create a new role for them: `oc create role rqviewer --verb=get,list,watch --resource=resourcequota,appliedclusterresourcequota -n ralpert-test`
5. Assign the role: `oc adm policy add-role-to-user rqviewer rebecca --role-namespace=ralpert-test`
6. Create a cluster resource quota:
```
oc create clusterquota for-user \
     --project-annotation-selector openshift.io/requester=rebecca \
     --hard pods=10 \
     --hard secrets=20
```
7. Change user to `rebecca` and run `oc get appliedclusterresourcequota`. If you can see an ACRQ, you should see it in the UI list.
8. The link should take you to a custom ACRQ details page.
